matrix/plugins/glitch/actions.py

```diff
@@ -0,0 +1,69 @@
+import asyncio
+import logging
+from typing import Any
+from functools import wraps
+
+import enforce
+from juju.model import Model
+from juju.unit import Unit
+
+from matrix.utils import Singleton
+
+log = logging.getLogger("glitch")
+
+
+class _Actions(dict, metaclass=Singleton):
+    """
+    Interface for registering actions, translating text strings into
+    action functions, and performing automagic around our action
+    functions.
+
+    """
+    def decorate(self, func):
+        """
+        Register an action. Return a function that accepts a set of objects,
+        and iterates over that set, running the registered action on each
+        object in that set.
+
+        """
+        @enforce.runtime_validation
+        @wraps(func)
+        async def wrapped(model, objects, **kwargs):
+            for obj in objects:
+                await func(model, obj, **kwargs)
+        self[func.__name__] = wrapped
+        return wrapped
+
+# Public singleton
+Actions = _Actions()
+action = Actions.decorate
+
+
+#
+# Define your actions here
+#
+@action
+async def reboot(model: Model, unit: Unit):
```

Comment:
Note for self: If an action can accept multiple types (e.g., units or applications), it can use `Union` (e.g., `Union[Unit, Application]`).